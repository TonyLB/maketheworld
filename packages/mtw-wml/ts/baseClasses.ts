import { ParseTag } from "./parser/baseClasses";

type SchemaNestingBase<T extends SchemaTag> = {
    contents: T[];
}

export type LegalAssetContents = SchemaActionTag | SchemaComputedTag | SchemaConditionTag | SchemaExitTag | SchemaFeatureTag | SchemaImageTag | SchemaImportTag | SchemaMapTag | SchemaRoomTag | SchemaVariableTag
export type LegalConditionContents = SchemaActionTag | SchemaComputedTag | SchemaConditionTag | SchemaExitTag | SchemaFeatureTag | SchemaImageTag | SchemaImportTag | SchemaMapTag | SchemaRoomTag | SchemaVariableTag

type SchemaAssetBase = {
    key: string;
    fileName?: string;
    zone?: string;
    subFolder?: string;
    player?: string;
} & SchemaNestingBase<LegalAssetContents>

export type SchemaAssetTag = {
    tag: 'Asset';
} & SchemaAssetBase

export type SchemaStoryTag = {
    tag: 'Story';
    instance: boolean;
} & SchemaAssetBase

export type SchemaImageTag = {
    tag: 'Image';
    key: string;
    fileURL?: string;
}

type SchemaPronouns = {
    subject: string;
    object: string;
    possessive: string;
    adjective: string;
    reflexive: string;
}

export type SchemaCharacterTag = {
    tag: 'Character';
    player?: string;
    name: string;
    pronouns: SchemaPronouns;
    firstImpression?: string;
    oneCoolThing?: string;
    outfit?: string;
    image?: SchemaImageTag;
}

export type SchemaVariableTag = {
    tag: 'Variable';
    key: string;
    default?: string;
}

export type SchemaComputedTag = {
    tag: 'Computed';
    key: string;
    src: string;
    dependencies: string[];
}

export type SchemaActionTag = {
    tag: 'Action';
    src: string;
}

export type SchemaUseTag = {
    tag: 'Use';
    key: string;
    as?: string;
    type?: string;
}

type SchemaImportMapping = {
    key: string;
    type: 'Room' | 'Feature' | 'Variable' | 'Computed' | 'Action' | 'Map'
}

export type SchemaImportTag = {
    tag: 'Import';
    from: string;
    mapping: Record<string, SchemaImportMapping>;
}

//
// TODO:  Figure out how to define limits on schemaConditionTag contents
// without causing circular reference
//
export type SchemaConditionTag = {
    tag: 'Condition';
    if: string;
    dependencies: string[];
} & SchemaNestingBase<any>

export type SchemaExitTag = {
    tag: 'Exit';
    key: string;
    name: string;
    to: string;
    from: string;
}

export type SchemaLinkTag = {
    tag: 'Link';
    to: string;
} & SchemaNestingBase<SchemaStringTag>

export type SchemaDescriptionTag = {
    tag: 'Description';
    display: 'before' | 'after' | 'replace';
    spaceBefore: boolean;
    spaceAfter: boolean;
} & SchemaNestingBase<SchemaStringTag | SchemaLinkTag | SchemaLineBreakTag>

export type SchemaLineBreakTag = {
    tag: 'br';
}

export type SchemaNameTag = {
    tag: 'Name';
    name: string;
}

export type SchemaRoomTag = {
    tag: 'Room';
    key: string;
    global: boolean;
    display?: string;
    x?: number;
    y?: number;
} & SchemaNestingBase<SchemaDescriptionTag | SchemaExitTag | SchemaFeatureTag | SchemaNameTag>

export type SchemaFeatureTag = {
    tag: 'Feature';
    key: string;
    global: boolean;
}& SchemaNestingBase<SchemaDescriptionTag | SchemaNameTag>

export type SchemaMapTag = {
    tag: 'Map';
    key: string;
} & SchemaNestingBase<SchemaExitTag | SchemaImageTag | SchemaRoomTag>

export type SchemaStringTag = {
    tag: 'String';
    value: string;
}

export type SchemaTag = SchemaAssetTag |
    SchemaStoryTag |
    SchemaCharacterTag |
    SchemaImageTag |
    SchemaVariableTag |
    SchemaComputedTag |
    SchemaActionTag |
    SchemaUseTag |
    SchemaImportTag |
    SchemaConditionTag |
    SchemaExitTag |
    SchemaDescriptionTag |
    SchemaLineBreakTag |
    SchemaLinkTag |
    SchemaNameTag |
    SchemaRoomTag |
    SchemaFeatureTag |
    SchemaMapTag |
    SchemaStringTag

export class SchemaException extends Error {
    parseTag: ParseTag;
    constructor(message: string, parseTag: ParseTag) {
        super(message)
        this.name = 'WMLSchemaException'
        this.parseTag = parseTag
    }
}

export const isSchemaString = (value: SchemaTag): value is SchemaStringTag => (value.tag === 'String')
