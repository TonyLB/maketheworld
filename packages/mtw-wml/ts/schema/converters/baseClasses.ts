import { SchemaTag } from "../baseClasses"
import { ParsePropertyTypes, ParseTagOpen, ParseTagSelfClosure } from "../../simpleParser/baseClasses"
import { GenericTree, GenericTreeNode } from "../../tree/baseClasses";

export type SchemaConverterArguments = {
    parseOpen: ParseTagOpen | ParseTagSelfClosure;
    contextStack: GenericTree<SchemaTag>;
}

export type SchemaInitialConverter = {
    (args: SchemaConverterArguments): SchemaTag
}

export type ValidationTemplateItem = {
    required?: boolean;
    type: ParsePropertyTypes;
}

export type ValidationTemplate = Record<string, ValidationTemplateItem>

export type ValidationTemplateRemap<V extends ValidationTemplate>  ={
    [K in keyof V]: V[K] extends { type: ParsePropertyTypes.Boolean } ? boolean : string
}

export type ValidationRequiredKeys<V extends ValidationTemplate> = {[K in keyof V]: V[K] extends { required: true } ? K : never}[keyof V]

export type ValidationTemplateOutput<V extends ValidationTemplate> = 
    Partial<ValidationTemplateRemap<V>> &
    Pick<ValidationTemplateRemap<V>, ValidationRequiredKeys<V>>

export type ConverterMapValidateProperties = {
    isValid: (item: SchemaTag) => boolean;
    branchTags: SchemaTag['tag'][];
    leafTags: SchemaTag['tag'][];
}

export type ConverterMapEntry = {
    initialize: SchemaInitialConverter;
    typeCheckContents?: (item: SchemaTag, contextStack: GenericTree<SchemaTag>) => boolean;
    validateContents?: ConverterMapValidateProperties;
    finalize?: (initialTag: SchemaTag, contents: GenericTree<SchemaTag>, contextStack: GenericTree<SchemaTag>) => GenericTreeNode<SchemaTag>;
    //
    // wrapper and aggregate are for tags that wrap their sibling groups together (e.g. If/ElseIf/Else). If wrapper is
    // present but aggregate is not then a wrapper is created. If aggregate is present then a previous wrapper must
    // already exist on contextStack, and the aggregate function is used to turn that wrapper object (together with
    // its current children) into the new wrapper object (with, presumably, updated children)
    //
    wrapper?: 'If';
    aggregate?: (previous: GenericTreeNode<SchemaTag>, node: GenericTreeNode<SchemaTag>) => GenericTreeNode<SchemaTag>;
}

export const isSchemaWrapper = (tag: SchemaTag['tag']): tag is 'If' => (['If'].includes(tag))

export type SchemaToWMLOptions = {
    indent: number;
    context: SchemaTag[];
    siblings?: GenericTree<SchemaTag>;
}

export enum PrintMapOptionsChange {
    Sibling,
    Indent
}
export type PrintMapOptionsFactory = {
    (action: PrintMapOptionsChange): (previous: SchemaToWMLOptions) => SchemaToWMLOptions;
}

export type PrintMapEntryArguments = {
    tag: GenericTreeNode<SchemaTag>;
    options: SchemaToWMLOptions;
    optionsFactory: PrintMapOptionsFactory;
    schemaToWML: PrintMapEntry;
}

export enum PrintMode {
    naive,
    nested,
    propertyNested
}

//
// prettyPrint functions will always return PrintMapResult. The following assumptions should apply:
//    - If printMode === naive then the output should be a single line without line breaks
//    - If printMode === nested then the output can be multiline, with the left-most elements
//      snug against the left margin (no global indent ... indent is applied recursively as each
//      parent element incorporates its children)
//    - If printMode === nested then the output can be either a single element or multiple
//      elements
//    - If printMode === propertyNested then the output can represent only a single element (and
//      its children)
//
export type PrintMapResult = {
    printMode: PrintMode;
    output: string;
}

export type PrintMapEntry = {
    (args: PrintMapEntryArguments): PrintMapResult[];
}
