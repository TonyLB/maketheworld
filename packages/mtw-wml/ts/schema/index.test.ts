import parse from '../parser'
import { ParseException } from '../parser/baseClasses'
import tokenizer from '../parser/tokenizer'
import SourceStream from '../parser/tokenizer/sourceStream'

import { schemaFromParse } from '.'

describe('schemaFromParse', () => {
    beforeEach(() => {
        jest.clearAllMocks()
        jest.resetAllMocks()
    })

    it('should make a schema from parse elements correctly', () => {
        const testParse = parse(tokenizer(new SourceStream(`
        <Asset key=(Test) fileName="test">
            <Import from=(BASE)>
                <Use key=(basePower) type="Variable" as=(power) />
                <Use key=(overview) type="Room" />
            </Import>
            <Room key=(ABC)>
                <Name>Vortex</Name>
                <Description>
                    <Space />
                    Vortex<If {open}>
                        : Open
                    </If><ElseIf {!closed}>
                        : Indeterminate
                    </ElseIf><Else>
                        : Closed
                    </Else>
                    <Link to=(toggleOpen)>(toggle)</Link>
                </Description>
                <Exit from=(DEF)>vortex</Exit>
            </Room>
            <If {open}>
                <Room key=(ABC)>
                    <Exit to=(DEF)>welcome</Exit>
                </Room>
            </If>
            <Room key=(DEF)>
                <Name>Welcome</Name>
            </Room>
            <Variable key=(open) default={false} />
            <Action key=(toggleOpen) src={open = !open} />
            <Computed key=(closed) src={!open} />
            <Moment key=(openDoorMoment)>
                <Message key=(openDoor)>
                    The door opens!
                    <Room key=(ABC) />
                </Message>
            </Moment>
        </Asset>
    `)))
        expect(schemaFromParse(testParse)).toMatchSnapshot()

    })

    it('should combine conditional elements at every level', () => {
        const testParse = parse(tokenizer(new SourceStream(`
        <Asset key=(Test) fileName="test">
            <Room key=(ABC)>
                <Description>
                    Test One
                    <If {open}>Test Two</If>
                </Description>
                <If {open}>
                    <Description>
                        Test Three
                    </Description>
                </If>
            </Room>
            <If {open}>
                <Room key=(ABC)>
                    <Description>Test Four</Description>
                </Room>
            </If>
            <Variable key=(open) default={false} />
        </Asset>
    `)))
        expect(schemaFromParse(testParse)).toMatchSnapshot()

    })

    it('should make a schema for a character correctly', () => {
        const testParse = parse(tokenizer(new SourceStream(`
        <Character key=(TESS) fileName="Tess" player="TonyLB">
        <Name>Tess</Name>
        <FirstImpression>Frumpy Goth</FirstImpression>
        <OneCoolThing>Fuchsia eyes</OneCoolThing>
        <Pronouns
            subject="she"
            object="her"
            possessive="her"
            adjective="hers"
            reflexive="herself"
        ></Pronouns>
        <Outfit>A bulky frock-coat lovingly kit-bashed from a black hoodie and patchily dyed lace.</Outfit>
        <Image key=(testIcon) />
    </Character>
    `)))
        expect(schemaFromParse(testParse)).toMatchSnapshot()

    })

    it('should correctly extract map rooms', () => {
        const testParse = parse(tokenizer(new SourceStream(`
        <Asset key=(Test) fileName="test">
            <Map key=(testMap)>
                <Room key=(ABC) x="100" y="0" />
                <If {open}>
                    <Room key=(DEF) x="-100" y="0" />
                </If>
            </Map>
            <Variable key=(open) default={false} />
        </Asset>
    `)))
        expect(schemaFromParse(testParse)).toMatchSnapshot()

    })

    it('should correctly schematize bookmarks', () => {
        const testParse = parse(tokenizer(new SourceStream(`
        <Story key=(Test) instance fileName="test">
            <Bookmark key=(postFix)>
                <Space />(awesome!)
            </Bookmark>
            <Room key=(ABC)>
                <Name>Vortex</Name>
                <Description>Vortex<Bookmark key=(postFix) /></Description>
            </Room>
        </Story>
    `)))
        expect(schemaFromParse(testParse)).toMatchSnapshot()

    })

})