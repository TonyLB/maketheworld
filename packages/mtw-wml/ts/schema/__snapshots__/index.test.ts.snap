// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`schemaFromParse should make a schema for a character correctly 1`] = `
Array [
  Object {
    "FirstImpression": undefined,
    "Name": "Tess",
    "OneCoolThing": "Fuchsia eyes",
    "Outfit": "A bulky frock-coat lovingly kit-bashed from a black hoodie and patchily dyed lace.",
    "Pronouns": Object {
      "adjective": "hers",
      "object": "her",
      "possessive": "her",
      "reflexive": "herself",
      "subject": "she",
    },
    "contents": Array [
      Object {
        "contents": Array [
          Object {
            "parse": Object {
              "endTagToken": 13,
              "startTagToken": 13,
              "tag": "String",
              "value": "Tess",
            },
            "tag": "String",
            "value": "Tess",
          },
        ],
        "parse": Object {
          "contents": Array [
            Object {
              "endTagToken": 13,
              "startTagToken": 13,
              "tag": "String",
              "value": "Tess",
            },
          ],
          "endTagToken": 14,
          "startTagToken": 11,
          "tag": "Name",
        },
        "tag": "Name",
      },
      Object {
        "contents": Array [
          Object {
            "parse": Object {
              "endTagToken": 21,
              "startTagToken": 21,
              "tag": "String",
              "value": "Fuchsia eyes",
            },
            "tag": "String",
            "value": "Fuchsia eyes",
          },
        ],
        "parse": Object {
          "contents": Array [
            Object {
              "endTagToken": 21,
              "startTagToken": 21,
              "tag": "String",
              "value": "Fuchsia eyes",
            },
          ],
          "endTagToken": 22,
          "startTagToken": 19,
          "tag": "OneCoolThing",
          "value": "Fuchsia eyes",
        },
        "tag": "OneCoolThing",
        "value": "Fuchsia eyes",
      },
      Object {
        "adjective": "hers",
        "object": "her",
        "parse": Object {
          "adjective": "hers",
          "endTagToken": 41,
          "object": "her",
          "possessive": "her",
          "reflexive": "herself",
          "startTagToken": 23,
          "subject": "she",
          "tag": "Pronouns",
        },
        "possessive": "her",
        "reflexive": "herself",
        "subject": "she",
        "tag": "Pronouns",
      },
      Object {
        "contents": Array [
          Object {
            "parse": Object {
              "endTagToken": 44,
              "startTagToken": 44,
              "tag": "String",
              "value": "A bulky frock-coat lovingly kit-bashed from a black hoodie and patchily dyed lace.",
            },
            "tag": "String",
            "value": "A bulky frock-coat lovingly kit-bashed from a black hoodie and patchily dyed lace.",
          },
        ],
        "parse": Object {
          "contents": Array [
            Object {
              "endTagToken": 44,
              "startTagToken": 44,
              "tag": "String",
              "value": "A bulky frock-coat lovingly kit-bashed from a black hoodie and patchily dyed lace.",
            },
          ],
          "endTagToken": 45,
          "startTagToken": 42,
          "tag": "Outfit",
          "value": "A bulky frock-coat lovingly kit-bashed from a black hoodie and patchily dyed lace.",
        },
        "tag": "Outfit",
        "value": "A bulky frock-coat lovingly kit-bashed from a black hoodie and patchily dyed lace.",
      },
      Object {
        "fileURL": "https://musique.opus-31.fr/images/aaa.png",
        "key": "testIcon",
        "parse": Object {
          "endTagToken": 54,
          "fileURL": "https://musique.opus-31.fr/images/aaa.png",
          "key": "testIcon",
          "startTagToken": 46,
          "tag": "Image",
        },
        "tag": "Image",
      },
    ],
    "fileName": "Tess",
    "fileURL": "https://musique.opus-31.fr/images/aaa.png",
    "key": "TESS",
    "parse": Object {
      "contents": Array [
        Object {
          "contents": Array [
            Object {
              "endTagToken": 13,
              "startTagToken": 13,
              "tag": "String",
              "value": "Tess",
            },
          ],
          "endTagToken": 14,
          "startTagToken": 11,
          "tag": "Name",
        },
        Object {
          "contents": Array [
            Object {
              "endTagToken": 21,
              "startTagToken": 21,
              "tag": "String",
              "value": "Fuchsia eyes",
            },
          ],
          "endTagToken": 22,
          "startTagToken": 19,
          "tag": "OneCoolThing",
          "value": "Fuchsia eyes",
        },
        Object {
          "adjective": "hers",
          "endTagToken": 41,
          "object": "her",
          "possessive": "her",
          "reflexive": "herself",
          "startTagToken": 23,
          "subject": "she",
          "tag": "Pronouns",
        },
        Object {
          "contents": Array [
            Object {
              "endTagToken": 44,
              "startTagToken": 44,
              "tag": "String",
              "value": "A bulky frock-coat lovingly kit-bashed from a black hoodie and patchily dyed lace.",
            },
          ],
          "endTagToken": 45,
          "startTagToken": 42,
          "tag": "Outfit",
          "value": "A bulky frock-coat lovingly kit-bashed from a black hoodie and patchily dyed lace.",
        },
        Object {
          "endTagToken": 54,
          "fileURL": "https://musique.opus-31.fr/images/aaa.png",
          "key": "testIcon",
          "startTagToken": 46,
          "tag": "Image",
        },
      ],
      "endTagToken": 55,
      "fileName": "Tess",
      "key": "TESS",
      "player": "TonyLB",
      "startTagToken": 0,
      "tag": "Character",
    },
    "player": "TonyLB",
    "subFolder": undefined,
    "tag": "Character",
    "zone": undefined,
  },
]
`;

exports[`schemaFromParse should make a schema from parse elements correctly 1`] = `
Array [
  Object {
    "Story": undefined,
    "contents": Array [
      Object {
        "from": "BASE",
        "mapping": Object {
          "overview": Object {
            "key": "overview",
            "type": "Room",
          },
          "power": Object {
            "key": "basePower",
            "type": "Variable",
          },
        },
        "parse": Object {
          "contents": Array [
            Object {
              "as": "power",
              "endTagToken": 24,
              "key": "basePower",
              "startTagToken": 13,
              "tag": "Use",
              "type": "Variable",
            },
            Object {
              "endTagToken": 33,
              "key": "overview",
              "startTagToken": 25,
              "tag": "Use",
              "type": "Room",
            },
          ],
          "endTagToken": 34,
          "from": "BASE",
          "startTagToken": 8,
          "tag": "Import",
        },
        "tag": "Import",
      },
      Object {
        "contents": Array [
          Object {
            "contents": Array [
              Object {
                "parse": Object {
                  "endTagToken": 49,
                  "startTagToken": 47,
                  "tag": "Space",
                },
                "tag": "Space",
              },
              Object {
                "parse": Object {
                  "endTagToken": 51,
                  "startTagToken": 51,
                  "tag": "String",
                  "value": "Vortex",
                },
                "tag": "String",
                "value": "Vortex",
              },
              Object {
                "parse": Object {
                  "contents": Array [
                    Object {
                      "endTagToken": 58,
                      "startTagToken": 58,
                      "tag": "String",
                      "value": "(toggle)",
                    },
                  ],
                  "endTagToken": 59,
                  "startTagToken": 53,
                  "tag": "Link",
                  "to": "toggleOpen",
                },
                "tag": "Link",
                "text": "(toggle)",
                "to": "toggleOpen",
              },
            ],
            "display": undefined,
            "parse": Object {
              "contents": Array [
                Object {
                  "endTagToken": 46,
                  "startTagToken": 46,
                  "tag": "Whitespace",
                },
                Object {
                  "endTagToken": 49,
                  "startTagToken": 47,
                  "tag": "Space",
                },
                Object {
                  "endTagToken": 50,
                  "startTagToken": 50,
                  "tag": "Whitespace",
                },
                Object {
                  "endTagToken": 51,
                  "startTagToken": 51,
                  "tag": "String",
                  "value": "Vortex",
                },
                Object {
                  "endTagToken": 52,
                  "startTagToken": 52,
                  "tag": "Whitespace",
                },
                Object {
                  "contents": Array [
                    Object {
                      "endTagToken": 58,
                      "startTagToken": 58,
                      "tag": "String",
                      "value": "(toggle)",
                    },
                  ],
                  "endTagToken": 59,
                  "startTagToken": 53,
                  "tag": "Link",
                  "to": "toggleOpen",
                },
                Object {
                  "endTagToken": 60,
                  "startTagToken": 60,
                  "tag": "Whitespace",
                },
              ],
              "endTagToken": 61,
              "startTagToken": 44,
              "tag": "Description",
            },
            "tag": "Description",
          },
          Object {
            "contents": Array [
              Object {
                "parse": Object {
                  "endTagToken": 67,
                  "startTagToken": 67,
                  "tag": "String",
                  "value": "vortex",
                },
                "tag": "String",
                "value": "vortex",
              },
            ],
            "from": "DEF",
            "key": "DEF#ABC",
            "name": "vortex",
            "parse": Object {
              "contents": Array [
                Object {
                  "endTagToken": 67,
                  "startTagToken": 67,
                  "tag": "String",
                  "value": "vortex",
                },
              ],
              "endTagToken": 68,
              "from": "DEF",
              "key": "DEF#ABC",
              "startTagToken": 62,
              "tag": "Exit",
              "to": "ABC",
            },
            "tag": "Exit",
            "to": "ABC",
          },
        ],
        "display": undefined,
        "global": undefined,
        "key": "ABC",
        "name": Array [
          Object {
            "parse": Object {
              "endTagToken": 42,
              "startTagToken": 42,
              "tag": "String",
              "value": "Vortex",
            },
            "tag": "String",
            "value": "Vortex",
          },
        ],
        "parse": Object {
          "contents": Array [
            Object {
              "contents": Array [
                Object {
                  "endTagToken": 42,
                  "startTagToken": 42,
                  "tag": "String",
                  "value": "Vortex",
                },
              ],
              "endTagToken": 43,
              "startTagToken": 40,
              "tag": "Name",
            },
            Object {
              "contents": Array [
                Object {
                  "endTagToken": 46,
                  "startTagToken": 46,
                  "tag": "Whitespace",
                },
                Object {
                  "endTagToken": 49,
                  "startTagToken": 47,
                  "tag": "Space",
                },
                Object {
                  "endTagToken": 50,
                  "startTagToken": 50,
                  "tag": "Whitespace",
                },
                Object {
                  "endTagToken": 51,
                  "startTagToken": 51,
                  "tag": "String",
                  "value": "Vortex",
                },
                Object {
                  "endTagToken": 52,
                  "startTagToken": 52,
                  "tag": "Whitespace",
                },
                Object {
                  "contents": Array [
                    Object {
                      "endTagToken": 58,
                      "startTagToken": 58,
                      "tag": "String",
                      "value": "(toggle)",
                    },
                  ],
                  "endTagToken": 59,
                  "startTagToken": 53,
                  "tag": "Link",
                  "to": "toggleOpen",
                },
                Object {
                  "endTagToken": 60,
                  "startTagToken": 60,
                  "tag": "Whitespace",
                },
              ],
              "endTagToken": 61,
              "startTagToken": 44,
              "tag": "Description",
            },
            Object {
              "contents": Array [
                Object {
                  "endTagToken": 67,
                  "startTagToken": 67,
                  "tag": "String",
                  "value": "vortex",
                },
              ],
              "endTagToken": 68,
              "from": "DEF",
              "key": "DEF#ABC",
              "startTagToken": 62,
              "tag": "Exit",
              "to": "ABC",
            },
          ],
          "endTagToken": 69,
          "key": "ABC",
          "startTagToken": 35,
          "tag": "Room",
          "x": undefined,
          "y": undefined,
        },
        "render": Array [
          Object {
            "parse": Object {
              "endTagToken": 49,
              "startTagToken": 47,
              "tag": "Space",
            },
            "tag": "Space",
          },
          Object {
            "parse": Object {
              "endTagToken": 51,
              "startTagToken": 51,
              "tag": "String",
              "value": "Vortex",
            },
            "tag": "String",
            "value": "Vortex",
          },
          Object {
            "parse": Object {
              "contents": Array [
                Object {
                  "endTagToken": 58,
                  "startTagToken": 58,
                  "tag": "String",
                  "value": "(toggle)",
                },
              ],
              "endTagToken": 59,
              "startTagToken": 53,
              "tag": "Link",
              "to": "toggleOpen",
            },
            "tag": "Link",
            "text": "(toggle)",
            "to": "toggleOpen",
          },
        ],
        "tag": "Room",
        "x": undefined,
        "y": undefined,
      },
      Object {
        "contents": Array [
          Object {
            "contents": Array [
              Object {
                "contents": Array [
                  Object {
                    "parse": Object {
                      "endTagToken": 91,
                      "startTagToken": 91,
                      "tag": "String",
                      "value": "welcome",
                    },
                    "tag": "String",
                    "value": "welcome",
                  },
                ],
                "from": "ABC",
                "key": "ABC#DEF",
                "name": "welcome",
                "parse": Object {
                  "contents": Array [
                    Object {
                      "endTagToken": 91,
                      "startTagToken": 91,
                      "tag": "String",
                      "value": "welcome",
                    },
                  ],
                  "endTagToken": 92,
                  "from": "ABC",
                  "key": "ABC#DEF",
                  "startTagToken": 86,
                  "tag": "Exit",
                  "to": "DEF",
                },
                "tag": "Exit",
                "to": "DEF",
              },
            ],
            "display": undefined,
            "global": undefined,
            "key": "ABC",
            "name": Array [],
            "parse": Object {
              "contents": Array [
                Object {
                  "contents": Array [
                    Object {
                      "endTagToken": 91,
                      "startTagToken": 91,
                      "tag": "String",
                      "value": "welcome",
                    },
                  ],
                  "endTagToken": 92,
                  "from": "ABC",
                  "key": "ABC#DEF",
                  "startTagToken": 86,
                  "tag": "Exit",
                  "to": "DEF",
                },
              ],
              "endTagToken": 93,
              "key": "ABC",
              "startTagToken": 81,
              "tag": "Room",
              "x": undefined,
              "y": undefined,
            },
            "render": Array [],
            "tag": "Room",
            "x": undefined,
            "y": undefined,
          },
        ],
        "dependencies": Array [
          "open",
        ],
        "if": "open",
        "parse": Object {
          "contents": Array [
            Object {
              "contents": Array [
                Object {
                  "contents": Array [
                    Object {
                      "endTagToken": 91,
                      "startTagToken": 91,
                      "tag": "String",
                      "value": "welcome",
                    },
                  ],
                  "endTagToken": 92,
                  "from": "ABC",
                  "key": "ABC#DEF",
                  "startTagToken": 86,
                  "tag": "Exit",
                  "to": "DEF",
                },
              ],
              "endTagToken": 93,
              "key": "ABC",
              "startTagToken": 81,
              "tag": "Room",
              "x": undefined,
              "y": undefined,
            },
          ],
          "dependencies": Array [
            Object {
              "endTagToken": 80,
              "on": "open",
              "startTagToken": 75,
              "tag": "Depend",
            },
          ],
          "endTagToken": 94,
          "if": "open",
          "startTagToken": 70,
          "tag": "Condition",
        },
        "tag": "Condition",
      },
      Object {
        "contents": Array [],
        "display": undefined,
        "global": undefined,
        "key": "DEF",
        "name": Array [
          Object {
            "parse": Object {
              "endTagToken": 102,
              "startTagToken": 102,
              "tag": "String",
              "value": "Welcome",
            },
            "tag": "String",
            "value": "Welcome",
          },
        ],
        "parse": Object {
          "contents": Array [
            Object {
              "contents": Array [
                Object {
                  "endTagToken": 102,
                  "startTagToken": 102,
                  "tag": "String",
                  "value": "Welcome",
                },
              ],
              "endTagToken": 103,
              "startTagToken": 100,
              "tag": "Name",
            },
          ],
          "endTagToken": 104,
          "key": "DEF",
          "startTagToken": 95,
          "tag": "Room",
          "x": undefined,
          "y": undefined,
        },
        "render": Array [],
        "tag": "Room",
        "x": undefined,
        "y": undefined,
      },
      Object {
        "default": "false",
        "key": "open",
        "parse": Object {
          "default": "false",
          "endTagToken": 113,
          "key": "open",
          "startTagToken": 105,
          "tag": "Variable",
        },
        "tag": "Variable",
      },
      Object {
        "key": "toggleOpen",
        "parse": Object {
          "endTagToken": 122,
          "key": "toggleOpen",
          "src": "open = !open",
          "startTagToken": 114,
          "tag": "Action",
        },
        "src": "open = !open",
        "tag": "Action",
      },
      Object {
        "dependencies": Array [
          "open",
        ],
        "key": "closed",
        "parse": Object {
          "dependencies": Array [
            Object {
              "endTagToken": 136,
              "on": "open",
              "startTagToken": 131,
              "tag": "Depend",
            },
          ],
          "endTagToken": 137,
          "key": "closed",
          "src": "!open",
          "startTagToken": 123,
          "tag": "Computed",
        },
        "src": "!open",
        "tag": "Computed",
      },
    ],
    "fileName": "test",
    "key": "Test",
    "parse": Object {
      "contents": Array [
        Object {
          "contents": Array [
            Object {
              "as": "power",
              "endTagToken": 24,
              "key": "basePower",
              "startTagToken": 13,
              "tag": "Use",
              "type": "Variable",
            },
            Object {
              "endTagToken": 33,
              "key": "overview",
              "startTagToken": 25,
              "tag": "Use",
              "type": "Room",
            },
          ],
          "endTagToken": 34,
          "from": "BASE",
          "startTagToken": 8,
          "tag": "Import",
        },
        Object {
          "contents": Array [
            Object {
              "contents": Array [
                Object {
                  "endTagToken": 42,
                  "startTagToken": 42,
                  "tag": "String",
                  "value": "Vortex",
                },
              ],
              "endTagToken": 43,
              "startTagToken": 40,
              "tag": "Name",
            },
            Object {
              "contents": Array [
                Object {
                  "endTagToken": 46,
                  "startTagToken": 46,
                  "tag": "Whitespace",
                },
                Object {
                  "endTagToken": 49,
                  "startTagToken": 47,
                  "tag": "Space",
                },
                Object {
                  "endTagToken": 50,
                  "startTagToken": 50,
                  "tag": "Whitespace",
                },
                Object {
                  "endTagToken": 51,
                  "startTagToken": 51,
                  "tag": "String",
                  "value": "Vortex",
                },
                Object {
                  "endTagToken": 52,
                  "startTagToken": 52,
                  "tag": "Whitespace",
                },
                Object {
                  "contents": Array [
                    Object {
                      "endTagToken": 58,
                      "startTagToken": 58,
                      "tag": "String",
                      "value": "(toggle)",
                    },
                  ],
                  "endTagToken": 59,
                  "startTagToken": 53,
                  "tag": "Link",
                  "to": "toggleOpen",
                },
                Object {
                  "endTagToken": 60,
                  "startTagToken": 60,
                  "tag": "Whitespace",
                },
              ],
              "endTagToken": 61,
              "startTagToken": 44,
              "tag": "Description",
            },
            Object {
              "contents": Array [
                Object {
                  "endTagToken": 67,
                  "startTagToken": 67,
                  "tag": "String",
                  "value": "vortex",
                },
              ],
              "endTagToken": 68,
              "from": "DEF",
              "key": "DEF#ABC",
              "startTagToken": 62,
              "tag": "Exit",
              "to": "ABC",
            },
          ],
          "endTagToken": 69,
          "key": "ABC",
          "startTagToken": 35,
          "tag": "Room",
          "x": undefined,
          "y": undefined,
        },
        Object {
          "contents": Array [
            Object {
              "contents": Array [
                Object {
                  "contents": Array [
                    Object {
                      "endTagToken": 91,
                      "startTagToken": 91,
                      "tag": "String",
                      "value": "welcome",
                    },
                  ],
                  "endTagToken": 92,
                  "from": "ABC",
                  "key": "ABC#DEF",
                  "startTagToken": 86,
                  "tag": "Exit",
                  "to": "DEF",
                },
              ],
              "endTagToken": 93,
              "key": "ABC",
              "startTagToken": 81,
              "tag": "Room",
              "x": undefined,
              "y": undefined,
            },
          ],
          "dependencies": Array [
            Object {
              "endTagToken": 80,
              "on": "open",
              "startTagToken": 75,
              "tag": "Depend",
            },
          ],
          "endTagToken": 94,
          "if": "open",
          "startTagToken": 70,
          "tag": "Condition",
        },
        Object {
          "contents": Array [
            Object {
              "contents": Array [
                Object {
                  "endTagToken": 102,
                  "startTagToken": 102,
                  "tag": "String",
                  "value": "Welcome",
                },
              ],
              "endTagToken": 103,
              "startTagToken": 100,
              "tag": "Name",
            },
          ],
          "endTagToken": 104,
          "key": "DEF",
          "startTagToken": 95,
          "tag": "Room",
          "x": undefined,
          "y": undefined,
        },
        Object {
          "default": "false",
          "endTagToken": 113,
          "key": "open",
          "startTagToken": 105,
          "tag": "Variable",
        },
        Object {
          "endTagToken": 122,
          "key": "toggleOpen",
          "src": "open = !open",
          "startTagToken": 114,
          "tag": "Action",
        },
        Object {
          "dependencies": Array [
            Object {
              "endTagToken": 136,
              "on": "open",
              "startTagToken": 131,
              "tag": "Depend",
            },
          ],
          "endTagToken": 137,
          "key": "closed",
          "src": "!open",
          "startTagToken": 123,
          "tag": "Computed",
        },
      ],
      "endTagToken": 138,
      "fileName": "test",
      "key": "Test",
      "startTagToken": 0,
      "tag": "Asset",
    },
    "player": undefined,
    "subFolder": undefined,
    "tag": "Asset",
    "zone": undefined,
  },
]
`;
